app/Platform/Actions/BuildGameListAction.php

```diff
@@ -0,0 +1,582 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Platform\Actions;
+
+use App\Community\Enums\AwardType;
+use App\Community\Enums\TicketState;
+use App\Community\Enums\UserGameListType;
+use App\Data\PaginatedData;
+use App\Models\Game;
+use App\Models\Leaderboard;
+use App\Models\PlayerGame;
+use App\Models\Ticket;
+use App\Models\User;
+use App\Platform\Data\GameData;
+use App\Platform\Data\GameListEntryData;
+use App\Platform\Data\PlayerGameData;
+use App\Platform\Enums\AchievementFlag;
+use App\Platform\Enums\GameListType;
+use App\Platform\Enums\UnlockMode;
+use Illuminate\Database\Eloquent\Builder;
+use Illuminate\Pagination\LengthAwarePaginator;
+use Illuminate\Support\Collection;
+use Illuminate\Support\Facades\DB;
+use InvalidArgumentException;
+
+class BuildGameListAction
+{
+    public function execute(
+        GameListType $listType,
+        ?User $user,
+        int $perPage = 25,
+        int $page = 1,
+        array $sort = [],
+        array $filters = [],
+    ): PaginatedData {
+        // Regardless of the list context, we'll build a common base query which can use
+        // the reusable sorts and filters and then be passed to a datatable component.
+        $query = $this->buildBaseQuery($listType, $user);
+
+        // After building the base query, tack on whatever filters and sort we need.
+        // We support multiple filters, but only a single selected sort.
+        $this->applyFilters($query, $filters, $user);
+        $this->applySorting($query, $sort, $user);
+
+        // The base query may not respect ->distinct() operations done by some filters.
+        // We'll override its `total` value with the correct one.
+        $total = $query->count('GameData.ID');
+
+        /** @var LengthAwarePaginator<Game> $entries */
+        $entries = $query->paginate($perPage, ['*'], 'page', $page);
```

Comment:
I'm not sure exactly where the difference is coming from, but this generates a massive query that takes 45.8ms to run (just the one query - I'm ignoring the additional 6.7ms to get the playerGames and backlogGames), whereas the equivalent queries on the Want to Play page take 12.98ms (2.58+4.63+0.31+3.33+2.13)

Reply:
It's coming from `BuildGameListAction::applyGameTitleSorting()`.

`GameListService` does the sorting in memory with PHP, whereas `BuildGameListAction` is doing it with SQL. The bottleneck can be proven out by adding a `return;` to the top of `applyGameTitleSorting()`. Doing so drops the query time on my machine to 2ms.

I have explored doing the sort in-memory in this action similar to how `GameListService` does it, but doing so implies that the sort needs to happen before pagination. I'm very concerned about the long-term scalability implications of doing this.

I think what might be the best approach is to compute `SortPrefix` and SortTitle` in advance and store them directly on the `GameData` table as new columns. I may tackle that in a separate PR.

Reply:
Consider #1062 before you tackle that.

Reply:
Thanks for sharing that.

Seems like it's almost the same thing. All games will have an autogenerated sort title, but it could manually be overwritten perhaps in the management app?

Reply:
`sort_title` implementation is in https://github.com/RetroAchievements/RAWeb/pull/2690.